plumpy/process_spec.py

```diff
@@ -171,15 +171,59 @@ def has_output(self, name):
         """
         return name in self.outputs
 
-    def validate_inputs(self, inputs=None):
+    def pre_process_inputs(self, port_namespace, inputs):
+        """
+        Take the passed input arguments and match it to the ports of the port namespace,
+        filling in any default values for inputs that have not been supplied as long as the
+        default is defined
+
+        :param port_namespace: the port namespace against which to compare the inputs dictionary
+        :param inputs: the dictionary with supplied inputs
+        :return: an AttributesFrozenDict with the inputs, complemented with port default values
+        :raises: ValueError if no input was specified for a required port without a default value
         """
-        Validate a dictionary of inputs according to the input port namespace of this specification
+        from . import utils
+
+        result = dict(inputs) if inputs else {}
+
+        for name, port in port_namespace.items():
+
+            if name not in inputs:
+                if port.has_default():
+                    port_value = port.default
+                elif port.required:
+                    raise ValueError('Value not supplied for required inputs port {}'.format(name))
+                elif isinstance(port, ports.PortNamespace):
+                    port_value = {}
+                else:
+                    continue
+            else:
+                port_value = inputs[name]
+
+            if isinstance(port, ports.PortNamespace):
+                result[name] = self.pre_process_inputs(port, port_value)
+            else:
+                result[name] = port_value
+
+        return utils.AttributesFrozendict(result)
+
+    def validate_inputs(self, inputs=None):
+        """Process and validate a dictionary of inputs according to the input port namespace of this specification.
+
+        The pre processing step will also process ports with defaults that did not receive an explicit value.
 
         :param inputs: the inputs dictionary
-        :return: valid or not, error string|None
-        :rtype: tuple(bool, str or None)
+        :return: the parsed inputs
+        :rtype: AttributeFrozenDict
+        :raise ValueError: if the inputs are invalid
         """
-        return self.inputs.validate(inputs)
+        inputs = self.pre_process_inputs(self.inputs, inputs or {})
+        result = self.inputs.validate(inputs)
```

Comment:
so, we have `inputs.validate` that does seem to return errors, if they exist
and `validate_inputs` which seems to return the inputs again.

Can this be homogenized?

Reply:
the `validate` is a method for any port to validate itself for a given mapping of values. In this case, it is is the `PortNamespace` for the inputs, returned by `ProcessSpec.inputs`. The process spec itself has multiple `PortNamespaces`, i.e. `self.outputs`, so its validation method name needs to be more specific. The reason for having the `ProcessSpec.validate_inputs` return the processed or "parsed" inputs, is because the `Process` who calls it, also needs those inputs. If I did not return them, I would have to parse them again, or I would have to parse them before and then pass them in. But I think this pre-processing should be as close as possible to the port namespace and thus the process spec, not the process.

Reply:
One question: Why not make the `pre_process_inputs` function a member of the port namespace, just like `validate`?
So, you could call
```python
input_values = self.inputs.pre_process(input_values)
self.inputs.validate(input_values)
```

As for `validate_inputs`: since this is intended to be called by a process, is this function really needed here in the spec?
It is anyhow a shorthand for "preprocess, validate and raise errors" - if needed it could go into the process, if not you could just move those lines to wherever they are needed.

Reply:
I like making the pre processing a member of the `PortNamespace`. You are right the `validate_inputs` is just a convenience wrapper to process and validate. This is currently only used in the process and the tests. The process can easily do it itself. It will just add some code duplication to the tests